02_bubble/willshen8/bubble.go

```diff
@@ -0,0 +1,68 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+)
+
+var out io.Writer = os.Stdout
+
+func bubble(s []int) []int {
+	sliceCopy := append(s[:0:0], s...)
+	sliceLength := len(s)
+	sorted := false
+
+	for !sorted {
+		swapOccurred := false
+		for i := 0; i < sliceLength-1; i++ {
+			if sliceCopy[i+1] < sliceCopy[i] {
+				swap(sliceCopy, i, i+1)
+				swapOccurred = true
+			}
+		}
+		if !swapOccurred {
+			sorted = true
+		}
+	}
+	return sliceCopy
+}
+
+func insertionSort(s []int) []int {
+	sliceCopy := append(s[:0:0], s...)
+	for i := 1; i < len(s); i++ {
+		currValue := sliceCopy[i]
+		emptySlot := i
+
+		for emptySlot > 0 && sliceCopy[emptySlot-1] > currValue {
```

Comment:
an alternative here is to use copy() to move the existing entries rather than shuffling them along one at a time.